Client certificates silently ignored if paths specified with "~/"

**Describe the bug**
Client certificate is ignored if path is specified with "~/...".

Basically:
```
[http "https://.../"]
       sslCert = ~/trt/sec/keystore_..._cert.pem # working in git, but ignored in git lfs
       sslKey =  ~/trt/sec/keystore_..._key.pem  # working in git, but ignored in git lfs
       sslCert = /home/dev/trt/sec/keystore_..._cert.pem # working in both
       sslKey =  /home/dev/trt/sec/keystore_..._key.pem  # working in both
```

I found these env and set them:
```
GIT_TRACE=1
GIT_TRACE_PACKET=1
GIT_TRACE_PERFORMANCE=1
GIT_TRACE_SETUP=1
GIT_CURL_VERBOSE=1
GIT_TRANSFER_TRACE=1
```
Running "git lfs pull" shows this for about a second:
`Downloading LFS objects:   0% (0/1), 0 B | 0 B/s`
and then this:
```
batch response: Client error: https://{url}/{group}/{project_name}.git/info/lfs/objects/batch
batch response: Client error: https://{url}/{group}/{project_name}.git/info/lfs/objects/batch
batch response: Client error: https://{url}/{group}/{project_name}.git/info/lfs/objects/batch
batch response: Client error: https://{url}/{group}/{project_name}.git/info/lfs/objects/batch
batch response: Client error: https://{url}/{group}/{project_name}.git/info/lfs/objects/batch
batch response: Client error: https://{url}/{group}/{project_name}.git/info/lfs/objects/batch
batch response: Client error: https://{url}/{group}/{project_name}.git/info/lfs/objects/batch
batch response: Client error: https://{url}/{group}/{project_name}.git/info/lfs/objects/batch
batch response: Client error: https://{url}/{group}/{project_name}.git/info/lfs/objects/batch
batch response: Client error: https://{url}/{group}/{project_name}.git/info/lfs/objects/batch
batch response: Client error: https://{url}/{group}/{project_name}.git/info/lfs/objects/batch
batch response: Client error: https://{url}/{group}/{project_name}.git/info/lfs/objects/batch
batch response: Client error: https://{url}/{group}/{project_name}.git/info/lfs/objects/batch
batch response: Client error: https://{url}/{group}/{project_name}.git/info/lfs/objects/batch
batch response: Client error: https://{url}/{group}/{project_name}.git/info/lfs/objects/batch
Failed to fetch some objects from 'https://{url}/{group}/{project_name}.git/info/lfs'
```

I had this problem for months but never found out why. That's why I disabled git lfs for normal "git pull"
and just tried to find out why it wasn't working from time to time.

Now after using the full path it's working as expected.

Encountered the problem in a docker container, where I wanted to use the same client certificate for the host and the container.
That was the reason, why I was not using a full path, but ~/ instead and this worked for a normal git pull.


**To Reproduce**
Steps to reproduce the behavior:
1. Set up client certificates with leading "~/".
2. Verify that client certificates are working with git pull
3. Try git lfs pull
4. See error

**Expected behavior**
git lfs using the client certificate and no error.
A error, if sslCert and/or sslKey is set and can't be loaded instead of silently ignoring it.

**System environment**
Host: manjaro
Docker container: Ubuntu "20.04.3 LTS (Focal Fossa)"

git config -l
```
git-lfs/3.4.1 (GitHub; linux amd64; go 1.21.1)
git version 2.43.0

Endpoint=https://{url}/{group}/{project_name}.git/info/lfs (auth=basic)
LocalWorkingDir=/workspaces/{project_name}
LocalGitDir=/workspaces/{project_name}/.git
LocalGitStorageDir=/workspaces/{project_name}/.git
LocalMediaDir=/workspaces/{project_name}/.git/lfs/objects
LocalReferenceDirs=
TempDir=/workspaces/{project_name}/.git/lfs/tmp
ConcurrentTransfers=8
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=/workspaces/{project_name}/.git/lfs
AccessDownload=basic
AccessUpload=basic
DownloadTransfers=basic,lfs-standalone-file,ssh
UploadTransfers=basic,lfs-standalone-file,ssh
GIT_EXEC_PATH=/usr/lib/git-core
git config filter.lfs.process = "git-lfs filter-process --skip"
git config filter.lfs.smudge = "git-lfs smudge --skip -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"
```


**Output of `git lfs env`**
```
git-lfs/3.4.1 (GitHub; linux amd64; go 1.21.1)
git version 2.43.0

Endpoint=https://{url}/{group}/{project_name}.git/info/lfs (auth=basic)
LocalWorkingDir=/workspaces/{project_name}
LocalGitDir=/workspaces/{project_name}/.git
LocalGitStorageDir=/workspaces/{project_name}/.git
LocalMediaDir=/workspaces/{project_name}/.git/lfs/objects
LocalReferenceDirs=
TempDir=/workspaces/{project_name}/.git/lfs/tmp
ConcurrentTransfers=8
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=/workspaces/{project_name}/.git/lfs
AccessDownload=basic
AccessUpload=basic
DownloadTransfers=basic,lfs-standalone-file,ssh
UploadTransfers=basic,lfs-standalone-file,ssh
GIT_EXEC_PATH=/usr/lib/git-core
git config filter.lfs.process = "git-lfs filter-process --skip"
git config filter.lfs.smudge = "git-lfs smudge --skip -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"
```

**Additional context**

While writing this I created a dummy file "foobar" > \~/dummy_file and used it for cert and key.
Pointing to it with "~/dummy_file" results in git pull:
`fatal: unable to access 'https://{url}/{group}/{project_name}.git/': error reading X.509 key or certificate file`
and in the above error with git lfs pull.
I found this interesting because, when I use the full path I get the same error for git pull:
`fatal: unable to access 'https://{url}/{group}/{project_name}.git/': error reading X.509 key or certificate file`
but for git lfs pull I get:
```
panic: runtime error: invalid memory address or nil pointer dereference                                                                                                
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x82f02d]

goroutine 53 [running]:
crypto/x509.IsEncryptedPEMBlock(...)
	/usr/local/go/src/crypto/x509/pem_decrypt.go:105
github.com/git-lfs/git-lfs/v3/lfshttp.getClientCertForHost(0xc000167f40, {0xc00009c308, 0x13})
	/tmp/docker_run/src/git-lfs/amd64/lfshttp/certs.go:87 +0x34d
github.com/git-lfs/git-lfs/v3/lfshttp.(*Client).Transport(0xc000167f40, 0xc00031c1b0, {0xc00030054c, 0x5})
	/tmp/docker_run/src/git-lfs/amd64/lfshttp/client.go:467 +0x465
github.com/git-lfs/git-lfs/v3/lfshttp.(*Client).HttpClient(0xc000167f40, 0xc00031c1b0, {0xc00030054c, 0x5})
	/tmp/docker_run/src/git-lfs/amd64/lfshttp/client.go:508 +0x1cd
github.com/git-lfs/git-lfs/v3/lfsapi.(*Client).doWithCreds(0xc00031a600, 0xc000390b00, {{0xb52de0, 0xc00059ef60}, 0xc00059ee40, 0xc00031cab0, 0xc00059f0b0}, {{0xc00030054c, 0x5}, {0xc0000285f0, ...}}, ...)
	/tmp/docker_run/src/git-lfs/amd64/lfsapi/auth.go:100 +0x1a7
github.com/git-lfs/git-lfs/v3/lfsapi.(*Client).doWithAuth(0xc00031a600, {0xc0003003ed, 0x6}, {{0xc00030054c, 0x5}, {0xc0000285f0, 0x45}}, 0xc000390b00, {0x0, 0x0, ...})
	/tmp/docker_run/src/git-lfs/amd64/lfsapi/auth.go:68 +0x1cb
github.com/git-lfs/git-lfs/v3/lfsapi.(*Client).DoWithAuth(0xc00031a600, {0xc0003003ed, 0x6}, {{0xc00030054c, 0x5}, {0xc0000285f0, 0x45}}, 0xc000390b00)
	/tmp/docker_run/src/git-lfs/amd64/lfsapi/auth.go:26 +0x6d
github.com/git-lfs/git-lfs/v3/lfsapi.(*Client).DoAPIRequestWithAuth(0xc00031a600, {0xc0003003ed, 0x6}, 0xc000390b00)
	/tmp/docker_run/src/git-lfs/amd64/lfsapi/auth.go:57 +0x134
github.com/git-lfs/git-lfs/v3/tq.(*tqClient).Batch(0xc0003740c0, {0xc0003003ed, 0x6}, 0xc000511080)
	/tmp/docker_run/src/git-lfs/amd64/tq/api.go:94 +0x3dc
github.com/git-lfs/git-lfs/v3/tq.Batch({0xb56a10, 0xc000133080}, 0x1, {0xc0003003ed, 0x6}, 0x1?, {0xc000134a80, 0x64, 0x64})
	/tmp/docker_run/src/git-lfs/amd64/tq/api.go:45 +0x2b3
github.com/git-lfs/git-lfs/v3/tq.(*TransferQueue).enqueueAndCollectRetriesFor(0xc00016e500, {0xc0000dd880, 0x64, 0x0?})
	/tmp/docker_run/src/git-lfs/amd64/tq/transfer_queue.go:581 +0x2fe
github.com/git-lfs/git-lfs/v3/tq.(*TransferQueue).collectBatches.func1()
	/tmp/docker_run/src/git-lfs/amd64/tq/transfer_queue.go:470 +0x85
created by github.com/git-lfs/git-lfs/v3/tq.(*TransferQueue).collectBatches in goroutine 43
	/tmp/docker_run/src/git-lfs/amd64/tq/transfer_queue.go:463 +0x29b
```
And that's the reason I think it's getting silently ignored.

--- a/lfshttp/certs.go
+++ b/lfshttp/certs.go
@@ -10,7 +10,9 @@ import (
 	"path/filepath"
 
 	"github.com/git-lfs/git-lfs/v3/config"
+	"github.com/git-lfs/git-lfs/v3/errors"
 	"github.com/git-lfs/git-lfs/v3/tools"
+	"github.com/git-lfs/git-lfs/v3/tr"
 	"github.com/rubyist/tracerx"
 )
 
@@ -68,19 +70,19 @@ func decryptPEMBlock(c *Client, block *pem.Block, path string, key []byte) ([]by
 
 // getClientCertForHost returns a client certificate for a specific host (which may
 // be "host:port" loaded from the gitconfig
-func getClientCertForHost(c *Client, host string) *tls.Certificate {
+func getClientCertForHost(c *Client, host string) (*tls.Certificate, error) {
 	hostSslKey, _ := c.uc.Get("http", fmt.Sprintf("https://%v/", host), "sslKey")
 	hostSslCert, _ := c.uc.Get("http", fmt.Sprintf("https://%v/", host), "sslCert")
 
 	cert, err := os.ReadFile(hostSslCert)
 	if err != nil {
 		tracerx.Printf("Error reading client cert file %q: %v", hostSslCert, err)
-		return nil
+		return nil, errors.Wrapf(err, tr.Tr.Get("Error reading client cert file %q", hostSslCert))
 	}
 	key, err := os.ReadFile(hostSslKey)
 	if err != nil {
 		tracerx.Printf("Error reading client key file %q: %v", hostSslKey, err)
-		return nil
+		return nil, errors.Wrapf(err, tr.Tr.Get("Error reading client key file %q", hostSslKey))
 	}
 
 	block, _ := pem.Decode(key)
@@ -88,16 +90,16 @@ func getClientCertForHost(c *Client, host string) *tls.Certificate {
 		key, err = decryptPEMBlock(c, block, hostSslKey, key)
 		if err != nil {
 			tracerx.Printf("Unable to decrypt client key file %q: %v", hostSslKey, err)
-			return nil
+			return nil, errors.Wrapf(err, tr.Tr.Get("Error reading client key file %q (not a PKCS#1 file?)", hostSslKey))
 		}
 	}
 
 	certobj, err := tls.X509KeyPair(cert, key)
 	if err != nil {
 		tracerx.Printf("Error reading client cert/key %v", err)
-		return nil
+		return nil, errors.Wrapf(err, tr.Tr.Get("Error reading client cert/key"))
 	}
-	return &certobj
+	return &certobj, nil
 }
 
 // getRootCAsForHost returns a certificate pool for that specific host (which may

--- a/lfshttp/certs.go
+++ b/lfshttp/certs.go
@@ -74,6 +74,16 @@ func getClientCertForHost(c *Client, host string) (*tls.Certificate, error) {
 	hostSslKey, _ := c.uc.Get("http", fmt.Sprintf("https://%v/", host), "sslKey")
 	hostSslCert, _ := c.uc.Get("http", fmt.Sprintf("https://%v/", host), "sslCert")
 
+	hostSslKey, err := tools.ExpandPath(hostSslKey, false)
+	if err != nil {
+		return nil, errors.Wrapf(err, tr.Tr.Get("Error resolving key path %q", hostSslKey))
+	}
+
+	hostSslCert, err = tools.ExpandPath(hostSslCert, false)
+	if err != nil {
+		return nil, errors.Wrapf(err, tr.Tr.Get("Error resolving cert path %q", hostSslCert))
+	}
+
 	cert, err := os.ReadFile(hostSslCert)
 	if err != nil {
 		tracerx.Printf("Error reading client cert file %q: %v", hostSslCert, err)

--- a/lfshttp/certs.go
+++ b/lfshttp/certs.go
@@ -86,6 +86,9 @@ func getClientCertForHost(c *Client, host string) (*tls.Certificate, error) {
 	}
 
 	block, _ := pem.Decode(key)
+	if block == nil {
+		return nil, errors.New(tr.Tr.Get("Error decoding PEM block from %q", hostSslKey))
+	}
 	if x509.IsEncryptedPEMBlock(block) {
 		key, err = decryptPEMBlock(c, block, hostSslKey, key)
 		if err != nil {

--- a/lfshttp/client.go
+++ b/lfshttp/client.go
@@ -470,7 +470,10 @@ func (c *Client) Transport(u *url.URL, access creds.AccessMode) (http.RoundTripp
 
 	if isClientCertEnabledForHost(c, host) {
 		tracerx.Printf("http: client cert for %s", host)
-		cert := getClientCertForHost(c, host)
+		cert, err := getClientCertForHost(c, host)
+		if err != nil {
+			return nil, err
+		}
 		if cert != nil {
 			tr.TLSClientConfig.Certificates = []tls.Certificate{*cert}
 			tr.TLSClientConfig.BuildNameToCertificate()

--- a/t/t-clone.sh
+++ b/t/t-clone.sh
@@ -220,9 +220,58 @@ begin_test "clone ClientCert"
 )
 end_test
 
+begin_test "clone ClientCert with homedir certs"
+(
+  set -e
+
+  # Windows triggers a credential helper problem.
+  if [ $IS_WINDOWS -eq 1 ]; then
+    exit 0
+  fi
+
+  reponame="test-cloneClientCert-homedir"
+
+  cp "$LFS_CLIENT_KEY_FILE" "$HOME/lfs-client-key-file"
+  cp "$LFS_CLIENT_CERT_FILE" "$HOME/lfs-client-cert-file"
+
+  git config --global http.$LFS_CLIENT_CERT_URL/.sslKey "~/lfs-client-key-file"
+  git config --global http.$LFS_CLIENT_CERT_URL/.sslCert "~/lfs-client-cert-file"
+
+  setup_remote_repo "$reponame"
+  clone_repo_clientcert "$reponame" "$reponame"
+  if [ $(grep -c "client-cert-mac-openssl" clone_client_cert.log) -gt 0 ]; then
+    echo "Skipping due to SSL client cert bug in Git"
+    exit 0
+  fi
+
+  git lfs track "*.dat" 2>&1 | tee track.log
+  grep "Tracking \"\*.dat\"" track.log
+
+  # generate some test data & commits with random LFS data
+  echo "[
+  {
+    \"CommitDate\":\"$(get_date -5d)\",
+    \"Files\":[
+      {\"Filename\":\"file1.dat\",\"Size\":100},
+      {\"Filename\":\"file2.dat\",\"Size\":75}]
+  },
+  {
+    \"CommitDate\":\"$(get_date -1d)\",
+    \"Files\":[
+      {\"Filename\":\"file3.dat\",\"Size\":30}]
+  }
+  ]" | lfstest-testutils addcommits
+
+  git push origin main
+)
+end_test
+
 begin_test "clone with flags"
 (
   set -e
+
+  git config --global http.$LFS_CLIENT_CERT_URL/.sslCert "$LFS_CLIENT_CERT_FILE"
+  git config --global http.$LFS_CLIENT_CERT_URL/.sslKey "$LFS_CLIENT_KEY_FILE"
 
   reponame="$(basename "$0" ".sh")-flags"
   setup_remote_repo "$reponame"

--- a/t/t-fetch.sh
+++ b/t/t-fetch.sh
@@ -298,6 +298,20 @@ begin_test "fetch with missing object"
 )
 end_test
 
+begin_test "fetch does not crash on empty key files"
+(
+  set -e
+  cd clone
+  rm -rf .git/lfs/objects
+
+  git config --local http.sslKey /dev/null
+  git config --local http.sslCert /dev/null
+
+  git lfs fetch origin main 2>&1 | tee fetch.log
+  grep "Error decoding PEM block" fetch.log
+)
+end_test
+
 begin_test "fetch-all"
 (
   set -e

--- a/t/testhelpers.sh
+++ b/t/testhelpers.sh
@@ -562,6 +562,8 @@ setup_creds() {
   write_creds_file "user:pass" "$CREDSDIR/127.0.0.1"
   write_creds_file ":pass" "$CREDSDIR/--$certpath"
   write_creds_file ":pass" "$CREDSDIR/--$keypath"
+  write_creds_file ":pass" "$CREDSDIR/--$homecertpath"
+  write_creds_file ":pass" "$CREDSDIR/--$homekeypath"
 }
 
 # setup initializes the clean, isolated environment for integration tests.
@@ -620,6 +622,8 @@ setup() {
   # setup the git credential password storage
   local certpath="$(echo "$LFS_CLIENT_CERT_FILE" | tr / -)"
   local keypath="$(echo "$LFS_CLIENT_KEY_FILE_ENCRYPTED" | tr / -)"
+  local homecertpath="$(echo "$TRASHDIR/home/lfs-client-cert-file" | tr / -)"
+  local homekeypath="$(echo "$TRASHDIR/home/lfs-client-key-file" | tr / -)"
   setup_creds
 
   echo "#"

